presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -114,11 +114,14 @@ public ExchangeClient(
         this.maxBufferBytes = Long.MIN_VALUE;
     }
 
-    public synchronized ExchangeClientStatus getStatus()
```

Comment:
I'm pretty sure this violates the Java memory model.  The fields like `bufferBytes` at the bottom must be read within a lock.

Maybe reorder the code so the `allClients` loop is first and then you take the synchronized lock for the rest of the code.